standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/TableOperationsHandler.java

```diff
@@ -0,0 +1,380 @@
+    }
+    return resp;
+  }
+
+  private static TableOperationsHandler<?>
+      ofCache(String opId, boolean shouldCancel) throws TException, IOException {
+    TableOperationsHandler<?> tableOp = null;
+    if (opId != null) {
+      tableOp = OPID_TO_HANDLER.get(opId);
+      if (tableOp == null && !shouldCancel) {
+        throw new MetaException("Couldn't find the drop operation by " + opId);
+      }
+      if (shouldCancel) {
+        if (tableOp != null) {
+          tableOp.cancelOperation();
+        } else {
+          tableOp = new TableOperationsHandler<>(opId) {
+            @Override
+            public TableOpResp toTableOpResp() throws TException {
+              TableOpResp resp = new TableOpResp(opId);
+              resp.setMessage("Drop operation has been canceled");
+              resp.setFinished(true);
+              return resp;
+            }
+          };
+        }
+      }
+    }
+    return tableOp;
+  }
+
+  public static <T extends TBase> TableOperationsHandler<T> offer(IHMSHandler handler, T req)
+      throws TException, IOException {
+    if (req instanceof DropTableRequest dropTableReq) {
+      TableOperationsHandler<T> tableOp = (TableOperationsHandler<T>)
+          ofCache(dropTableReq.getId(), dropTableReq.isCancel());
+      if (tableOp == null) {
+        tableOp = new TableOperationsHandler<>(handler, dropTableReq.isAsyncDrop(), req);
+      }
+      return tableOp;
+    }
+    throw new UnsupportedOperationException("Not yet implemented");
+  }
+
+  private void runTableOperation() throws TException, IOException {
+    try {
+      // currently this class only supports the drop table, we can the alter table if possible in the future
+      if (request instanceof DropTableRequest dropReq) {
+        // drop any partitions
+        String catName = normalizeIdentifier(
+            dropReq.isSetCatalogName() ? dropReq.getCatalogName() : getDefaultCatalog(handler.getConf()));
+        String name = normalizeIdentifier(dropReq.getTableName());
+        String dbname = normalizeIdentifier(dropReq.getDbName());
+        tableName = new TableName(catName, dbname, name);
+        logMsgPrefix = "Drop " + id + " on table " + tableName + ":";
+        GetTableRequest req = new GetTableRequest(dropReq.getDbName(), dropReq.getTableName());
+        req.setCatName(catName);
+        tbl = handler.get_table_core(req);
+        if (tbl == null) {
+          throw new NoSuchObjectException(tableName + " doesn't exist");
+        }
+        if (tbl.getSd() == null) {
+          throw new MetaException("Table metadata is corrupted");
+        }
+        boolean tableDataShouldBeDeleted = checkTableDataShouldBeDeleted(tbl, dropReq.isDeleteData());
+        if (tbl.getSd().getLocation() != null) {
+          tblPath = new Path(tbl.getSd().getLocation());
+        }
+        if (tableDataShouldBeDeleted && tblPath != null) {
+          if (!handler.getWh().isWritable(tblPath.getParent())) {
```

Comment:
Also need check if `tblPath` is writable.

Reply:
This is the old behavior of drop table, for the`tableDataShouldBeDeleted` table, it should be either the acid or the old "managed" table. I guess the old assumes metastore user has the writable permission on table's location.
Do we see the permission issue on the table's directory?